src/main/java/core/basesyntax/ArrayList.java

```diff
@@ -1,48 +1,151 @@
+    }
+
+    private void increaseArray(List<T> list) {
+        Object[] tempList = new Object[list.size()];
+        for (int i = 0; i < list.size(); i++) {
+            tempList[i] = list.get(i);
+        }
+        Object[] temp = defaultArray.clone();
+        defaultArray = new Object[newCapacity];
+        System.arraycopy(temp, DEFAULT_INDEX, defaultArray, DEFAULT_INDEX, sizeCounter);
+        System.arraycopy(tempList, DEFAULT_INDEX, defaultArray, sizeCounter, tempList.length);
+        sizeCounter += list.size();
+    }
 
+    public int grow() {
+        return newCapacity + (newCapacity / 2);
+    }
+
+    private void rangeChecker(int index) {
+        if ((index < 0 || index >= sizeCounter) && index != DEFAULT_INDEX) {
+            throw new ArrayListIndexOutOfBoundsException("Out of bounds");
+        }
+    }
+
+    private void rangeCheckerForAdding(int index) {
+        if ((index < 0 || index > sizeCounter) && index != DEFAULT_INDEX) {
+            throw new ArrayListIndexOutOfBoundsException("Out of bounds");
+        }
     }
 
     @Override
     public void addAll(List<T> list) {
-
+        if (sizeCounter + list.size() > newCapacity) {
+            newCapacity = grow();
+        }
+        increaseArray(list);
     }
 
     @Override
     public T get(int index) {
-        return null;
+        rangeChecker(index);
+        return (T)defaultArray[index];
     }
 
     @Override
     public void set(T value, int index) {
-
+        rangeChecker(index);
+        if (index == sizeCounter) {
+            sizeCounter++;
+        }
+        defaultArray[index] = value;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        rangeChecker(index);
+        if (index == sizeCounter - 1) {
+            Object[] temp = defaultArray.clone();
+            defaultArray = new Object[newCapacity];
+            System.arraycopy(temp, DEFAULT_INDEX, defaultArray, DEFAULT_INDEX, index);
+            sizeCounter--;
+            return (T)temp[index];
+        }
+        Object[] temp = defaultArray.clone();
+        defaultArray = new Object[newCapacity];
+        System.arraycopy(temp, DEFAULT_INDEX, defaultArray, DEFAULT_INDEX, index);
+        System.arraycopy(temp, index + 1, defaultArray, index, sizeCounter);
+        sizeCounter--;
+        return (T)temp[index];
     }
 
     @Override
     public T remove(T element) {
-        return null;
+        for (int i = 0; i <= sizeCounter; i++) {
+            if (element == null) {
+                sizeCounter--;
+                return null;
+            }
+            if (element.equals(defaultArray[i])) {
+                Object[] temp = defaultArray.clone();
+                defaultArray = new Object[newCapacity];
+                System.arraycopy(temp, DEFAULT_INDEX, defaultArray, DEFAULT_INDEX, i);
+                System.arraycopy(temp, i + 1, defaultArray, i, sizeCounter);
```

Comment:
Can we just create a new empty array and copy the value from the previous one there?